r2-core/src/main/java/com/linkedin/r2/message/RequestContext.java

```diff
@@ -54,7 +54,9 @@ public RequestContext()
    */
   public RequestContext(RequestContext other)
   {
-    _localAttrs = Collections.synchronizedMap(new HashMap<>(other._localAttrs));
+    synchronized (other._localAttrs) {
+      _localAttrs = Collections.synchronizedMap(new HashMap<>(other._localAttrs));
+    }
   }
 
   private RequestContext(Map<String, Object> localAttrs)
```

Comment:
For safety/clarity I think it would be better if the incoming map here was also wrapped in a synchronizedMap, and that behavior removed from the clone method. Unfortunately we can't rely on the type system to enforce that _localAttrs is a synchronized map.

Reply:
I don't think that will help, since if the other map is not a SynchronizedMap then it can still be directly modified during the copy, since the caller has direct access to the underlying map. Instead I propose changing the constructor RequestContext(Map localAttrs) to also wrap the _localAttrs in a SynchronizedMap. Then, since the constructors are the only places _localAttrs can be set it is provable it is synchronized.

Reply:
Sorry if my suggestion wasn't clear - what you propose is the same as my suggestion :)